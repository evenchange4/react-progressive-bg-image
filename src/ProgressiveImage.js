import PropTypes from 'prop-types';
import compose from 'recompose/compose';
import setDisplayName from 'recompose/setDisplayName';
import defaultProps from 'recompose/defaultProps';
import setPropTypes from 'recompose/setPropTypes';
import mapPropsStream from 'recompose/mapPropsStream';
import { Observable } from 'rxjs/Observable';
import { async } from 'rxjs/scheduler/async';
import 'rxjs/add/operator/combineLatest';
import 'rxjs/add/operator/pluck';
import 'rxjs/add/operator/startWith';
import 'rxjs/add/operator/switchMapTo';
import 'rxjs/add/operator/delay';
import 'rxjs/add/operator/merge';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/mapTo';
import 'rxjs/add/operator/switch';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/observable/from';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/merge';
import Img from './Img';
import loadImage from './loadImage';

export const DELAY = 200;

export function ownerPropsToChildProps(
  propStream, // ownerProps
  imagePromise = loadImage, // :: string => promise => ({ src, isCached })
  t = DELAY, // delay in milliseconds
  scheduler = async, // rx scheduler
) {
  const props$ = Observable.from(propStream);
  const placeholder$ = props$.pluck('placeholder');
  const imagePromise$ = props$
    .pluck('src')
    .switchMap(imagePromise)
    .startWith({ src: '', isCached: false });

  const src$ = imagePromise$.pluck('src').filter(src => !!src);
  const isCached$ = imagePromise$.pluck('isCached').distinctUntilChanged();

  const isLoaded$ = Observable.merge(
    placeholder$.mapTo(Observable.of(false)),
    imagePromise$.map(({ isCached }) =>
      Observable.of(true).delay(isCached ? 0 : t, scheduler),
    ),
  )
    .switch()
    .startWith(false)
    .distinctUntilChanged();

  const image$ = placeholder$.merge(src$).distinctUntilChanged();

  return props$.combineLatest(
    image$,
    isCached$,
    isLoaded$,
    (props, image, isCached, isLoaded) => ({
      ...props,
      image,
      isCached,
      isLoaded,
    }),
  );
}

export default compose(
  setDisplayName('ProgressiveImage'),
  setPropTypes({
    src: PropTypes.string.isRequired,
    placeholder: PropTypes.string.isRequired,
    opacity: PropTypes.number,
    blur: PropTypes.number,
    scale: PropTypes.number,
    component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  }),
  defaultProps({
    opacity: 0.5,
    blur: 20,
    scale: 1,
    transition: 'opacity 0.3s linear',
    component: 'div',
  }),
  mapPropsStream(ownerPropsToChildProps),
)(Img);
