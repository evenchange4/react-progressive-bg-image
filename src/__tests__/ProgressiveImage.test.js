import React from 'react';
import { mount } from 'enzyme';
import ProgressiveImage from '../ProgressiveImage';

it('should render <ProgressiveImage>', () => {
  const wrapper = mount(
    <ProgressiveImage
      src="origin.jpg"
      placeholder="small.jpg"
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
