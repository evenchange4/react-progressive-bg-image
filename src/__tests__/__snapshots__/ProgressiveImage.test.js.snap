// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. before resolving image 1`] = `
<ProgressiveImage
  blur={20}
  component="div"
  opacity={0.5}
  placeholder="small.jpg"
  scale={1}
  src="origin.jpg"
  transition="opacity 0.3s linear"
>
  <mapPropsStream(Img)
    blur={20}
    component="div"
    opacity={0.5}
    placeholder="small.jpg"
    scale={1}
    src="origin.jpg"
    transition="opacity 0.3s linear"
  >
    <Img
      blur={20}
      component="div"
      image="small.jpg"
      isCached={false}
      isLoaded={false}
      opacity={0.5}
      placeholder="small.jpg"
      scale={1}
      src="origin.jpg"
      transition="opacity 0.3s linear"
    >
      <Styled(BaseComponent)
        blur={20}
        component="div"
        isCached={false}
        isLoaded={false}
        opacity={0.5}
        placeholder="small.jpg"
        scale={1}
        src="origin.jpg"
        style={
          Object {
            "backgroundImage": "url(\\"small.jpg\\")",
          }
        }
        transition="opacity 0.3s linear"
      >
        <BaseComponent
          blur={20}
          className="sc-bdVaJa lpfHmJ"
          component="div"
          isCached={false}
          isLoaded={false}
          opacity={0.5}
          placeholder="small.jpg"
          scale={1}
          src="origin.jpg"
          style={
            Object {
              "backgroundImage": "url(\\"small.jpg\\")",
            }
          }
          transition="opacity 0.3s linear"
        >
          <div
            className="sc-bdVaJa lpfHmJ"
            src="origin.jpg"
            style={
              Object {
                "backgroundImage": "url(\\"small.jpg\\")",
              }
            }
          />
        </BaseComponent>
      </Styled(BaseComponent)>
    </Img>
  </mapPropsStream(Img)>
</ProgressiveImage>
`;

exports[`2. after resolving image 1`] = `
<ProgressiveImage
  blur={20}
  component="div"
  opacity={0.5}
  placeholder="small.jpg"
  scale={1}
  src="origin.jpg"
  transition="opacity 0.3s linear"
>
  <mapPropsStream(Img)
    blur={20}
    component="div"
    opacity={0.5}
    placeholder="small.jpg"
    scale={1}
    src="origin.jpg"
    transition="opacity 0.3s linear"
  >
    <Img
      blur={20}
      component="div"
      image="small.jpg"
      isCached={false}
      isLoaded={false}
      opacity={0.5}
      placeholder="small.jpg"
      scale={1}
      src="origin.jpg"
      transition="opacity 0.3s linear"
    >
      <Styled(BaseComponent)
        blur={20}
        component="div"
        isCached={false}
        isLoaded={false}
        opacity={0.5}
        placeholder="small.jpg"
        scale={1}
        src="origin.jpg"
        style={
          Object {
            "backgroundImage": "url(\\"small.jpg\\")",
          }
        }
        transition="opacity 0.3s linear"
      >
        <BaseComponent
          blur={20}
          className="sc-bdVaJa lpfHmJ"
          component="div"
          isCached={false}
          isLoaded={false}
          opacity={0.5}
          placeholder="small.jpg"
          scale={1}
          src="origin.jpg"
          style={
            Object {
              "backgroundImage": "url(\\"small.jpg\\")",
            }
          }
          transition="opacity 0.3s linear"
        >
          <div
            className="sc-bdVaJa lpfHmJ"
            src="origin.jpg"
            style={
              Object {
                "backgroundImage": "url(\\"small.jpg\\")",
              }
            }
          />
        </BaseComponent>
      </Styled(BaseComponent)>
    </Img>
  </mapPropsStream(Img)>
</ProgressiveImage>
`;
